packages/ember-data/tests/integration/records/property-changes-test.js

```diff
@@ -0,0 +1,115 @@
+var env, store, Person;
+var attr = DS.attr;
+var run = Ember.run;
+
+module('integration/records/property-changes - Property changes', {
+  setup: function() {
+    Person = DS.Model.extend({
+      firstName: attr('string'),
+      lastName: attr('string')
+    });
+    Person.toString = function() { return 'Person'; };
+
+    env = setupStore({
+      person: Person
+    });
+    store = env.store;
+  },
+
+  teardown: function() {
+    Ember.run(function() {
+      env.container.destroy();
+    });
+  }
+});
+
+test('Calling push with partial records trigger observers for just those attributes that changed', function() {
+  expect(1);
+  var person;
+
+  run(function() {
+    person = store.push('person', {
+      id: 'wat',
+      firstName: 'Yehuda',
+      lastName: 'Katz'
+    });
+  });
+
+  person.addObserver('firstName', function() {
+    ok(false, 'firstName observer should not be triggered');
+  });
+
+  person.addObserver('lastName', function() {
+    ok(true, 'lastName observer should be triggered');
+  });
+
+  run(function() {
+    store.push('person', {
+      id: 'wat',
+      firstName: 'Yehuda',
+      lastName: 'Katz!'
+    });
+  });
+});
+
+test('Calling push does not trigger observers for locally changed attributes with the same value', function() {
+  expect(0);
+  var person;
+
+  run(function() {
+    person = store.push('person', {
+      id: 'wat',
+      firstName: 'Yehuda',
+      lastName: 'Katz'
+    });
+
+    person.set('lastName', 'Katz!');
+  });
+
+  person.addObserver('firstName', function() {
+    ok(false, 'firstName observer should not be triggered');
+  });
+
+  person.addObserver('lastName', function() {
+    ok(false, 'lastName observer should not be triggered');
+  });
+
+  run(function() {
+    store.push('person', {
+      id: 'wat',
+      firstName: 'Yehuda',
+      lastName: 'Katz!'
+    });
+  });
+});
+
+test('Saving a record trigger observers for locally changed attributes with the same canonical value', function() {
+  expect(1);
+  var person;
+
+  env.adapter.updateRecord = function(store, type, snapshot) {
+    return Ember.RSVP.resolve({ id: 'wat', lastName: 'Katz' });
+  };
+
+  run(function() {
+    person = store.push('person', {
+      id: 'wat',
+      firstName: 'Yehuda',
+      lastName: 'Katz'
+    });
+
+    person.set('lastName', 'Katz!');
+  });
+
+  person.addObserver('firstName', function() {
+    ok(false, 'firstName observer should not be triggered');
+  });
+
+  person.addObserver('lastName', function() {
+    ok(true, 'lastName observer should be triggered');
+  });
+
+  run(function() {
+    person.save();
+  });
+});
```

Comment:
That's a good question... thinking about it a bit more I _think_ this is how we want it:

Pushing should only update canonical values (`_data`). If an attribute is present in `_attributes` or `_inFlightAttributes` we should not trigger, no matter what. If an attribute is present only in `_data` and the value differs from the new value, then we should trigger.

Saving should work more or less the same way but only take `_data` and `_inFlightAttributes` into consideration. If the attribute is present in `_attributes` we should not trigger.

Does this make sense?